src/python/pants/engine/target.py

```diff
@@ -501,5 +504,51 @@ def has_sources(self) -> bool:
     return SourcesResult(snapshot)
 
 
+# TODO: move this conversion into the engine. Create a singleton to access all the registered
+#  targets (or maybe just use unions). Consider having that singleton be a dict from alias ->
+#  Target for efficient lookups (we're guaranteed only one alias per target type).
+def hydrated_struct_to_target(
+    hydrated_struct: HydratedStruct, *, target_types: Iterable[Type[Target]]
+) -> Target:
+    kwargs = hydrated_struct.value.kwargs().copy()
+    type_alias = kwargs.pop("type_alias")
+
+    # We special case `address` and the field `name`. The `Target` constructor requires an
+    # `Address`, so we use the value pre-calculated via `build_files.py`'s `hydrate_struct` rule.
+    # We throw away the field `name` because it can be accessed via `tgt.address.target_name`, so
+    # there is no (known) reason to preserve the field.
+    # TODO: is this safe to do..? We annotate HydratedStruct.value.address as Optional. Are we
```

Comment:
As it stands, all `Structs` in `pants.engine.legacy` are `Targets`... anything that isn't a `Struct` is an "object" (in `BuildFileAliases` terminology), and those are not addressable.

Reply:
So is this an overly-strict annotation on `HydratedStruct.value.address` then?